### PR TITLE
Add composition for MariaDB service

### PR DESCRIPTION
**Pull request by TheBigLee**
## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.
- [ ] Link this PR to related issues or PRs.

### CODE REVIEW

File: component/component/vshn_appcat_services.jsonnet [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+local getServiceNamePlural(serviceName) =
+  local serviceNameLower = std.asciiLower(serviceName);
+  if std.endsWith(serviceName, 's') then
+    serviceNameLower
+  else
+    serviceNameLower + 's';
+
+local vshn_appcat_service(name) =
+  local isOpenshift = std.startsWith(inv.parameters.facts.distribution, 'openshift');
+  local securityContext = if isOpenshift then false else true;
```
Comment (Kidswiss) [inline component/component/vshn_appcat_services.jsonnet:29]: Isn't `isOpenshift` already a boolean? Can't you just do something like `!isOpenshift`?

File: component/component/vshn_appcat_services.jsonnet [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    subjects_: [ restoreServiceAccount ],
+  };
+
+  local resizeServiceAccount = kube.ServiceAccount('sa-sts-deleter') + {
+    metadata+: {
+      namespace: params.services.controlNamespace,
+    },
+  };
+
+  local resizeClusterRole = kube.ClusterRole('appcat:job:' + name + ':resizejob') {
```
Comment (Kidswiss) [inline component/component/vshn_appcat_services.jsonnet:60]: Wouldn't it make more sense to just have one set of clusterRoles and service accounts for all helm based services? The way this is currently implemented they all get the same hardcoded clusterRoles anyway.

File: component/component/vshn_mariadb.jsonnet [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+local mariadbPlans = common.FilterDisabledParams(mariadbParams.plans);
+
+local xrd = xrds.XRDFromCRD(
+  'xvshnminios.vshn.appcat.vshn.io',
+  xrds.LoadCRD('vshn.appcat.vshn.io_vshnminios.yaml', params.images.appcat.tag),
+  defaultComposition='vshnminio.vshn.appcat.vshn.io',
+  connectionSecretKeys=connectionSecretKeys,
+) + xrds.WithPlanDefaults(mariadbPlans, mariadbParams.defaultPlan);
+
+local composition =
```
Comment (Kidswiss) [inline component/component/vshn_mariadb.jsonnet:42]: Why aren't you using your shiny new `vshn_appcat_service()` here?

File: component/tests/vshn.yml [hunk 4]
```diff
@@ -55,9 +55,16 @@ parameters:
         emailAlerting:
           enabled: true
           smtpPassword: "whatever"
+        services:
+          mariadb:
+            enabled: true
+            grpcEndpoint: 172.18.0.1:9443
```
Comment (Kidswiss) [inline component/tests/vshn.yml:61]: Don't forget to comment out in the end. Although the productive deployment won't actually look at this.

File: docs/modules/ROOT/pages/runbooks/vshn-mariadb.adoc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+* A ticket alert means that the error rate is slightly higher than the objective.
+If we don't intervene at some point after receiving this alert, we can expect to miss the objective.
+However, no immediate, urgent action is necessary.
+A ticket alert should have a label `severity: warning`.
+* A page alert means that the error rate is significantly higher than the objective.
+Immediate action is necessary to not miss the objective.
+
+=== icon:bug[] Steps for debugging
+
+Failed probes can have a multitude of reasons, but in general there are two different kinds of issue clases.
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-mariadb.adoc:29]:
```suggestion
Failed probes can have a multitude of reasons, but in general there are two different kinds of issue classes.
```

File: docs/modules/ROOT/pages/runbooks/vshn-mariadb.adoc [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+To figure out what went wrong with provisioning it usually helps to take a closer look at the composite.
+
+[source,shell]
+----
+kubectl --as cluster-admin describe xvshnmariadb $COMPOSITE
+----
+
+If there are sync issues there usually are events that point to the root cause of the issue.
+
+Further it can help to look at the `Object` resources that are created for this instance.
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-mariadb.adoc:60]: Maybe also mention the `Release` object, as that is basically the core of the whole service, so it's worth checking as well :D

File: docs/modules/ROOT/pages/runbooks/vshn-minio.adoc [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+To figure out what went wrong with provisioning it usually helps to take a closer look at the composite.
+
+[source,shell]
+----
+kubectl --as cluster-admin describe xvshnminio $COMPOSITE
+----
+
+If there are sync issues there usually are events that point to the root cause of the issue.
+
+Further it can help to look at the `Object` resources that are created for this instance.
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-minio.adoc:60]: Here as well.

File: docs/modules/ROOT/pages/runbooks/vshn-redis.adoc [hunk 8]
```diff
@@ -0,0 +1,118 @@
+= Redis by VSHN
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-redis.adoc:1]: It feels like 80% of these runbooks could be put in their own `adoc` file and then included. I think asciidoc also support some variable injection and stuff.

Then we'd have the generic `is everything synced and ready` steps in one place and only have to document the service specific for each service.

File: tests/e2e/mariadb/01-connect.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
+apiVersion: batch/v1
```
Comment (Kidswiss) [inline tests/e2e/mariadb/01-connect.yaml:1]: I like that we test with a connect pod, but I think for future tests we could also check for specifics like:
* are there quota annotations on the namespace
* does the resizing work and adjust the pvcs
* is the claim SYNCED/HEALTHY -> I would actually add this assert now already, because that one will catch a lot of stuff right out of the bat

**Review by Kidswiss — CHANGES_REQUESTED** [review]
I like the generalizations so far 👍 

Also, the e2e tests are a good start, but I would really also add a check if the claim gets SYNCED/HEALTHY, as this will catch a lot of things that can go wrong. All the other suggestions for the tests are for future ideas.

File: component/class/defaults.yml [hunk 10]
```diff
@@ -39,7 +42,7 @@ parameters:
       appcat:
         registry: ghcr.io
         repository: vshn/appcat
-        tag: v4.43.1
+        tag: fix/missing_rbac_mariadb
```
Comment (zugao) [inline component/class/defaults.yml:45]: Don't forget to change that

File: tests/e2e/mariadb/01-connect.yaml (hunk 9 above)
Comment (TheBigLee) [reply tests/e2e/mariadb/01-connect.yaml:1]: I'm still in the middle of working for the e2e tests. I just commited what I aleady had so I don't loose work.
The whole e2e stuff will be worked on today.

File: component/class/defaults.yml [hunk 11]
```diff
@@ -64,7 +67,54 @@ parameters:
     tenantID: ${cluster:tenant}
     quotasEnabled: false
     grpcEndpoint: host.docker.internal:9443
-
+    defaultRestoreRoleRules:
+      - apiGroups:
+          - vshn.appcat.vshn.io
+        resources:
+          - vshnredis
```
Comment (zugao) [inline component/class/defaults.yml:74]: why only vshnredis here?

**Review by zugao — CHANGES_REQUESTED** [review]
This PR adds much more than what is in the title, I would have split it in 2 PRs. It's not clear which code is for E2E and which is just the composition and everything related to that.

File: component/component/helm_monitoring.jsonnet [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+local inv = kap.inventory();
+local params = inv.parameters.appcat;
+
+
+local helmMonitoringClusterRole = kube.ClusterRole('allow-helm-monitoring-resources') {
+  rules: [
+    {
+      apiGroups: [ 'monitoring.coreos.com' ],
+      resources: [ 'servicemonitors' ],
+      verbs: [ '*' ],
```
[the remark this answers is not in the document]
Comment (TheBigLee) [reply component/component/helm_monitoring.jsonnet:13]: Removed this whole file as it's not used after all

File: component/component/provider.jsonnet [hunk 13]
```diff
@@ -285,6 +285,11 @@ local runtimeConfigRef(name) = {
           resources: [ 'jobs' ],
           verbs: [ 'get', 'list', 'watch', 'create', 'delete' ],
         },
+        {
+          apiGroups: [ 'monitoring.coreos.com' ],
+          resources: [ 'servicemonitors' ],
+          verbs: [ '*' ],
```
[the remark this answers is not in the document]
Comment (TheBigLee) [reply component/component/provider.jsonnet:291]: Fixed

File: component/component/main.jsonnet [hunk 14]
```diff
@@ -167,10 +167,14 @@ local emailSecret = kube.Secret(params.services.vshn.emailAlerting.secretName) {
   '10_appcat_maintenance_recording_rule': maintenanceRule,
   [if params.services.vshn.enabled && params.services.vshn.emailAlerting.enabled then '10_mailgun_secret']: emailSecret,
 
-} + if params.slos.enabled then {
-  [if params.services.vshn.enabled && params.services.vshn.postgres.enabled then 'sli_exporter/90_slo_vshn_postgresql']: slos.Get('vshn-postgresql'),
-  [if params.services.vshn.enabled && params.services.vshn.postgres.enabled then 'sli_exporter/90_slo_vshn_postgresql_ha']: slos.Get('vshn-postgresql-ha'),
-  [if params.services.vshn.enabled && params.services.vshn.redis.enabled then 'sli_exporter/90_slo_vshn_redis']: slos.Get('vshn-redis'),
-  [if params.services.vshn.enabled && params.services.vshn.redis.enabled then 'sli_exporter/90_slo_vshn_redis_ha']: slos.Get('vshn-redis-ha'),
+} + if params.slos.enabled && params.services.vshn.enabled then {
```
[the remark this answers is not in the document]
Comment (TheBigLee) [reply component/component/main.jsonnet:170]: File has been removed. Not relevant anymore

File: component/component/vshn_appcat_services.jsonnet (hunk 1 above)
Comment (TheBigLee) [reply component/component/vshn_appcat_services.jsonnet]: yeah makes sense. Fixed

File: component/component/vshn_appcat_services.jsonnet (hunk 2 above)
Comment (TheBigLee) [reply component/component/vshn_appcat_services.jsonnet:60]: Removed it from this file as well as the redis file (to avoid duplicate manifests) and moved it to the sts-resize-controller file.

File: component/component/vshn_mariadb.jsonnet (hunk 3 above)
Comment (TheBigLee) [reply component/component/vshn_mariadb.jsonnet:42]: Forgot to remove this file before I started with the vshn_appcat_service()

File: component/tests/vshn.yml (hunk 4 above)
Comment (TheBigLee) [reply component/tests/vshn.yml:61]: This has already been refactored accordingly

File: docs/modules/ROOT/pages/runbooks/vshn-redis.adoc (hunk 8 above)
Comment (TheBigLee) [reply docs/modules/ROOT/pages/runbooks/vshn-redis.adoc:1]: Refactored the docs

File: docs/modules/ROOT/pages/runbooks/vshn-minio.adoc (hunk 7 above)
Comment (TheBigLee) [reply docs/modules/ROOT/pages/runbooks/vshn-minio.adoc]: fixed

File: component/component/vshn_appcat_services.jsonnet [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+local getServiceNamePlural(serviceName) =
+  local serviceNameLower = std.asciiLower(serviceName);
+  if std.endsWith(serviceName, 's') then
+    serviceNameLower
+  else
+    serviceNameLower + 's';
+
+local vshn_appcat_service(name) =
+  local isOpenshift = std.startsWith(inv.parameters.facts.distribution, 'openshift');
+  local securityContext = !isOpenshift;
```
Comment (Kidswiss) [inline component/component/vshn_appcat_services.jsonnet:29]: This seems now unused.

File: component/component/statefuleset-resize-controller.jsonnet [hunk 16]
```diff
@@ -66,10 +66,47 @@ local deployment = loadManifest('config/manager/manager.yaml') + {
...
+};
+
+local resizeClusterRoleBinding = kube.ClusterRoleBinding('appcat:job:resizejob') + {
+  roleRef_: resizeClusterRole,
+  subjects_: [ resizeServiceAccount ],
+};
+
 // Curently we only need this for redis.
-if params.services.vshn.enabled && params.services.vshn.redis.enabled then {
+if params.services.vshn.enabled && params.services.vshn.stsResizer.enabled then {
```
Comment (Kidswiss) [inline component/component/statefuleset-resize-controller.jsonnet:106]: Why did you change this? The stsResizer is not really usable on its own, it should be automatically installed when services are enabled that depend on it, otherwise it's yet another config flag we actively need to think about. It's already complex enough to set all the right flags to enable a service, let's not increase the configuration complexity if not necessary.

**Review by Kidswiss — CHANGES_REQUESTED** [review]
Something to look at for the sts resizer.

File: component/component/statefuleset-resize-controller.jsonnet (hunk 16 above)
Comment (TheBigLee) [reply component/component/statefuleset-resize-controller.jsonnet:106]: Switched it back to the original.
However, this appproach is also not verify feasable in the future, as we would have to include every future helm service in this if clause in the future.

File: component/class/defaults.yml (hunk 10 above)
Comment (TheBigLee) [reply component/class/defaults.yml]: Done
